Library/Homebrew/download_strategy.rb

```diff
@@ -282,6 +283,14 @@ def initialize(name, resource)
   end
 
   def fetch
+    if ENV["HOMEBREW_ARTIFACT_DOMAIN"]
+      if @url.start_with?(BottleSpecification::DEFAULT_DOMAIN)
+        @url.sub!(BottleSpecification::DEFAULT_DOMAIN, ENV["HOMEBREW_ARTIFACT_DOMAIN"])
```

Comment:
Handling bottles are easy as its coming from only one source which is ```https://homebrew.bintray.com/bottles``` but for non-bottle we don't know from where to fetch it until we read URL from Formula definition.
what we do underneath is  
for ex  when a user says ```brew install riak```
we get the archive of riak from ```https://s3.amazonaws.com/downloads.basho.com/riak/2.1/2.1.4/osx/10.8/riak-2.1.4-OSX-x86_64.tar.gz``` and store it in homebrew repository.